File: ui/src/ipc/meetings.ts
/**
 * Thin IPC client for the meeting-list + meeting-open surface (Phase 4, FR-33).
 *
 * The backend commands (`list_meetings`, `open_meeting`, `rename_meeting`,
 * `delete_meeting`, `reprocess`) are present on the generated `commands`
 * surface (Stream C / the Phase-4 backend JOIN regenerated `bindings.ts`). This
 * module reaches them through the shim-aware `commands` from `./client` (NOT raw
 * `./bindings`) — the same single injection point `notes.ts` uses, so the DEV
 * shim and the Vitest mocks both intercept here.
 *
 * Re-summarisation now goes through the Phase-5 `../ipc/summary` seam
 * (`summarise_meeting`): the meeting-list row's Summarise action runs the real
 * summariser via the summary store, so the Phase-4 `re_summarise` stub (and its
 * `reSummarise` wrapper here) were removed.
 *
 * Tests mock THIS module (per `architecture/cross-cutting.md` — Automated
 * testing policy); they do not fake the generated bindings file.
 *
 * `MeetingListEntry` and `MeetingState` are re-exported from the generated
 * `./bindings` (the canonical `common::MeetingListEntry` / `common::MeetingState`
 * shapes) so call sites and the DEV shim keep importing them from this seam with
 * no change.
 */
import { commands, unwrap } from "./client";
import type { MeetingId } from "./bindings";
import type { MeetingListEntry, MeetingState } from "./bindings";

export type { MeetingListEntry, MeetingState };

// The Phase-4 meeting commands are exposed on the shim-aware `commands` surface
// from `./client`. This module wraps them so callers work with plain
// `MeetingListEntry[]` / `MeetingState` / `void` results and so tests mock THIS
// module rather than the generated bindings file.

/** List all meetings for the meeting-list view (FR-33). */
export async function listMeetings(): Promise<MeetingListEntry[]> {
  return unwrap(await commands.listMeetings());
}

/** Open a meeting, returning its full restorable state. */
export async function openMeeting(meetingId: MeetingId): Promise<MeetingState> {
  return unwrap(await commands.openMeeting(meetingId));
}

/** Rename a meeting. */
export async function renameMeeting(
  meetingId: MeetingId,
  title: string,
): Promise<void> {
  unwrap(await commands.renameMeeting(meetingId, title));
}

/**
 * Set a speaker's display name (maps a diarizer label such as `"A"` to a
 * name); an empty `name` clears it. Returns the updated label→name map.
 */
export async function setSpeakerName(
  meetingId: MeetingId,
  label: string,
  name: string,
): Promise<Partial<Record<string, string>>> {
  return unwrap(await commands.setSpeakerName(meetingId, label, name));
}

/** Delete a meeting and its on-disk folder. */
export async function deleteMeeting(meetingId: MeetingId): Promise<void> {
  unwrap(await commands.deleteMeeting(meetingId));
}

/**
 * Open a meeting's on-disk directory in the host OS file explorer (themed
 * context menus, #0034 — meeting-list "Open storage folder" entry). The
 * backend resolves the path server-side and hands it to the platform opener;
 * no filesystem path crosses the IPC boundary.
 */
export async function openMeetingFolder(meetingId: MeetingId): Promise<void> {
  unwrap(await commands.openMeetingFolder(meetingId));
}

/**
 * Reprocess a meeting offline (#0015): re-transcribe THEN re-diarize under one
 * offline claim. Merges the former `reTranscribe` (FR-33) + `rediarize` (FR-11)
 * seams into a single command.
 *
 * Tests mock THIS module (per `architecture/cross-cutting.md` — Automated
 * testing policy); they do not fake the generated bindings file.
 *
 * The backend (`Orchestrator::reprocess`) re-runs ASR over the complete
 * `audio.opus` (rewriting `transcript.json`), then runs the `SherpaDiarizer`
 * over the FRESH transcript and finalises once — updating `metadata.json`'s
 * `speaker_count` + `diarizer`, refreshing the index row, and emitting
 * `AppEvent::TranscriptSegment` + `AppEvent::DiarizationComplete` on the shared
 * bus. The diarize step clears `metadata.json`'s `speaker_names` (re-lettering
 * can change who each label is), so any user-assigned speaker names are reset.
 * Long-running (ASR + sherpa inference on `spawn_blocking` tasks); resolves once
 * the whole pass completes. Refuses unless the recorder is `Idle`.
 */
export async function reprocess(meetingId: MeetingId): Promise<void> {
  unwrap(await commands.reprocess(meetingId));
}

/**
 * Voiceprint correction path (#0003 §2.4): clear the uncertain-band
 * auto-assigned name for `label` on `meetingId` and remove that
 * meeting/label's embedding contribution from the identity's gallery so it
 * does not bias future identifications.
 */
export async function rejectMatch(
  meetingId: MeetingId,
  label: string,
  identityId: string,
  modelId: string,
): Promise<void> {
  unwrap(await commands.rejectMatch(meetingId, label, identityId, modelId));
}

File: crates/ipc-bridge/src/commands/meetings.rs
//! Meeting list, open, and per-meeting action commands (Phase 4).
use super::*;


/// List all meetings for the meeting-list view (FR-33), most-recent first.
///
/// Reads straight from the libsql `index.db` ([`MeetingIndex::list_meetings`])
/// — a cheap projection that never loads a meeting's full transcript. The index
/// is async (libsql/tokio); the future is awaited here, never `block_on`'d.
#[tauri::command]
#[specta::specta]
pub async fn list_meetings(state: State<'_, IpcState>) -> AppResult<Vec<MeetingListEntry>> {
    // Self-heal: index any meeting present on disk but missing from the cache
    // (e.g. the process killed between finalise and the stop-time upsert) so it
    // can never stay hidden until the next startup `rebuild_from_disk`. Cheap
    // (a readdir + set-diff; only NEW folders are read). Best-effort — a
    // reconcile failure must not break listing, so log and serve the cache.
    if let Err(e) = state.index.reconcile_orphans(&state.meetings_dir).await {
        tracing::warn!(
            target: "ipc-bridge",
            "meeting-list self-heal reconcile failed: {e}; listing cached rows as-is"
        );
    }
    state.index.list_meetings().await
}

/// Open a meeting, returning its full restorable [`MeetingState`]
/// (metadata + transcript + optional notes).
///
/// Resolves the per-meeting folder under `IpcState::meetings_dir` and assembles
/// the state via `persistence::read_meeting_state`. The blocking `std::fs` reads
/// run on `spawn_blocking` per the threading model — the index is not consulted
/// (the folder is authoritative for the full state).
#[tauri::command]
#[specta::specta]
pub async fn open_meeting(
    meeting_id: MeetingId,
    state: State<'_, IpcState>,
) -> AppResult<MeetingState> {
    let meetings_dir = state.meetings_dir.clone();
    tokio::task::spawn_blocking(move || open_meeting_inner(&meetings_dir, meeting_id))
        .await
        .map_err(|e| AppError::Internal {
            context: format!("open_meeting task join failed: {e}"),
        })?
}

/// Rename a meeting: updates `metadata.json` (authoritative) then the index row.
///
/// Routes to `persistence::meeting_ops::rename_meeting`, which keeps the on-disk
/// folder and the index consistent (see `architecture/components.md`,
/// `persistence` "Phase 4 surface growth — meeting ops").
#[tauri::command]
#[specta::specta]
pub async fn rename_meeting(
    meeting_id: MeetingId,
    title: String,
    state: State<'_, IpcState>,
) -> AppResult<()> {
    meeting_ops::rename_meeting(&state.meetings_dir, &state.index, meeting_id, &title)
        .await
}

/// Set a speaker's display name on a saved meeting.
///
/// Maps a diarizer label (e.g. `"A"`) to a display name in `metadata.json`'s
/// `speaker_names`; an empty `name` clears the mapping. Returns the updated
/// map so the caller can re-render without re-reading the meeting. Names are
/// keyed by the diarizer label, so re-running diarization (which re-letters
/// speakers) resets them. Routes to `persistence::meeting_ops::set_speaker_name`.
///
/// The label and name are each capped at `MAX_SPEAKER_NAME_LEN` characters so
/// the UI cannot persist an unbounded value (mirrors the `set_speaker_name`
/// chat tool's bound).
#[tauri::command]
#[specta::specta]
pub async fn set_speaker_name(
    meeting_id: MeetingId,
    label: String,
    name: String,
    state: State<'_, IpcState>,
) -> AppResult<std::collections::BTreeMap<String, String>> {
    const MAX_SPEAKER_NAME_LEN: usize = 512;
    let name = name.trim().to_string();
    if label.chars().count() > MAX_SPEAKER_NAME_LEN || name.chars().count() > MAX_SPEAKER_NAME_LEN {
        return Err(minutist_common::AppError::InvalidInput {
            context: format!("speaker label/name too long (max {MAX_SPEAKER_NAME_LEN} characters)"),
        });
    }
    let result = meeting_ops::set_speaker_name(&state.meetings_dir, meeting_id, &label, &name)
        .await?;

    // Best-effort voiceprint enrolment: gate on the settings flag and a live
    // VoiceprintStore. Errors are logged and swallowed so a rename never fails
    // because of an enrolment problem (§2.3 asymmetry — accepted and documented).
    // The MCP/agent-tools set_speaker_name path (tools.rs) does NOT enrol:
    // it has no audio/diarizer access (§2.3 path asymmetry, intended).
    if state.settings.current().voiceprint_enrolment_enabled {
        if let Some(store) = state.voiceprints.as_ref().as_ref() {
            match state
                .orchestrator
                .enrol_voiceprint(meeting_id, label.clone(), name.clone(), store)
                .await
            {
                Ok(Some(id)) => {
                    tracing::debug!(
                        target: "ipc-bridge",
                        meeting_id = %meeting_id.0,
                        label = %label,
                        identity_id = %id.0,
                        "voiceprint enrolled after speaker rename"
                    );
                }
                Ok(None) => {
                    tracing::debug!(
                        target: "ipc-bridge",
                        meeting_id = %meeting_id.0,
                        label = %label,
                        "voiceprint enrolment skipped (no clean windows or model unavailable)"
                    );
                }
                Err(e) => {
                    tracing::warn!(
                        target: "ipc-bridge",
                        meeting_id = %meeting_id.0,
                        label = %label,
                        "voiceprint enrolment failed (best-effort, ignoring): {e}"
                    );
                }
            }
        }
    }

    Ok(result)
}

/// Move a meeting to the trash: it stays fully recoverable — the folder,
/// voiceprint contributions, and blobs are all left untouched — until
/// [`restore_meeting`] brings it back or [`purge_meeting`] (manual or the
/// 7-day auto-purge sweep) removes it for good.
///
/// Routes to `persistence::meeting_ops::soft_delete_meeting`.
#[tauri::command]
#[specta::specta]
pub async fn delete_meeting(
    meeting_id: MeetingId,
    state: State<'_, IpcState>,
) -> AppResult<()> {
    meeting_ops::soft_delete_meeting(
        &state.meetings_dir,
        &state.index,
        meeting_id,
        state.connected.sync.host_ref().await,
    )
    .await
}

/// Restore a meeting out of the trash — the mirror image of [`delete_meeting`].
///
/// Routes to `persistence::meeting_ops::restore_meeting`.
#[tauri::command]
#[specta::specta]
pub async fn restore_meeting(
    meeting_id: MeetingId,
    state: State<'_, IpcState>,
) -> AppResult<()> {
    meeting_ops::restore_meeting(
        &state.meetings_dir,
        &state.index,
        meeting_id,
        state.connected.sync.host_ref().await,
    )
    .await
}

/// Permanently remove a meeting ("Delete forever" on a trashed row): the
/// folder, its index row, its voiceprint contributions, and its blobs — then
/// records a purged tombstone so a hub replica can never resurrect it from a
/// slow peer (see `persistence::purged`). Unlike a soft delete, this is NOT
/// reversible.
///
/// The voiceprint purge is best-effort: if the `VoiceprintStore` is not open
/// (degraded-to-off) the step is skipped silently. The folder/index deletion
/// runs first so a crash between the two steps leaves at most an orphaned
/// voiceprint entry, not an orphaned meeting folder.
///
/// Routes to `persistence::meeting_ops::purge_meeting`.
#[tauri::command]
#[specta::specta]
pub async fn purge_meeting(
    meeting_id: MeetingId,
    state: State<'_, IpcState>,
) -> AppResult<()> {
    meeting_ops::purge_meeting(
        &state.meetings_dir,
        app_data_root(&state),
        &state.index,
        state.voiceprints.as_ref().as_ref(),
        meeting_id,
    )
    .await?;

    // Best-effort: unpin this meeting's blobs from the local blob store (a no-op
    // on the free build, or before the sync engine has started).
    if let Err(e) = state.connected.sync.delete_meeting_blobs(meeting_id).await {
        tracing::warn!(
            target: "ipc-bridge",
            meeting_id = %meeting_id.0,
            "unpinning purged meeting's blobs failed (best-effort): {e}"
        );
    }

    Ok(())
}

/// Open a meeting's on-disk directory (`{meetings_dir}/{uuid}/`) in the host
/// OS file explorer.
///
/// Mirrors [`open_attachment`]'s host hand-off: the path is resolved
/// server-side from `meeting_id` alone (never crosses the IPC boundary) and
/// handed to `tauri-plugin-opener`'s Rust API, so no opener capability scope
/// is required. The existence check runs on `spawn_blocking`; an absent
/// directory is `AppError::InvalidInput` rather than a silent no-op, since
/// calling the opener on a missing path would surface as a confusing OS-level
/// error instead of an app-level one.
#[tauri::command]
#[specta::specta]
pub async fn open_meeting_folder(
    meeting_id: MeetingId,
    app: tauri::AppHandle,
    state: State<'_, IpcState>,
) -> AppResult<()> {
    let meetings_dir = state.meetings_dir.clone();
    let path = tokio::task::spawn_blocking(move || -> Result<std::path::PathBuf, AppError> {
        let dir = meetings_dir.join(meeting_id.0.to_string());
        if !dir.is_dir() {
            return Err(AppError::InvalidInput {
                context: format!("meeting {meeting_id:?} has no on-disk directory"),
            });
        }
        Ok(dir)
    })
    .await
    .map_err(|e| AppError::Internal {
        context: format!("open_meeting_folder task join failed: {e}"),
    })??;

    use tauri_plugin_opener::OpenerExt;
    app.opener()
        .open_path(path.to_string_lossy().into_owned(), None::<&str>)
        .map_err(|e| AppError::Internal {
            context: format!("opening the meeting folder in the host file explorer failed: {e}"),
        })?;
    Ok(())
}

